KAFKA-5316: LogCleaner should account for larger record sets after cleaning

hachikuji opened the pull request.

hachikuji committed:
KAFKA-5316: LogCleaner should account for larger record sets after cleaning

hachikuji commented:
I'm aware of the checkstyle issues and will fix them in an upcoming patch. 

An alternative if we are concerned about the loss of the `writeOriginalEntry` optimization is to allocate the `ByteBufferOutputStream` at the start of `filterTo` and pass it to the `MemoryRecordsBuilder` instance.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -192,9 +179,9 @@ private static FilterResult filterTo(Iterable<MutableRecordBatch> batches, Recor
-                destinationBuffer.position(destinationBuffer.position() + slice.position());
                 messagesRetained += retainedRecords.size();
                 bytesRetained += records.sizeInBytes();
+                outputBuffer = builder.buffer();
```
If (1) outputBatch.size > inputBatch.size and (2) outputBatch.size > MaxMessageSize, perhaps we could log a warning that consumers before 0.10.2 may need to increase the fetch size?

junrao reviewed (commented):
@hachikuji : Thanks for finding the issue and providing the patch. The new logic seems simpler. LGTM. Just one minor comment.

hachikuji commented:
@junrao Thanks for the review. Something else that occurred to me is that the recovery logic validates that messages do not exceed the max message size (they are considered corrupt if they do). With this change, we will need to loosen that checking. Perhaps we could ensure that messages are not more than twice the max message size or something like that?

hachikuji committed:
Address review comments and fix checkstyle issues

hachikuji committed:
Allow record batches to exceed max message size when recovering log and in DumpLogSegments

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -139,72 +154,75 @@ private static FilterResult filterTo(Iterable<MutableRecordBatch> batches, Recor
                 }
+
+                // If we had to allocate a new buffer to fit the filtered output, reset the initial position
+                // in the original buffer to ensure that the partial data cannot mistakenly read. We also return
```
cannot mistakenly read => cannot mistakenly be read ?

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -264,7 +264,11 @@ class LogSegment(val log: FileRecords,
-      for (batch <- log.batches(maxMessageSize).asScala) {
+      // record batches can grow during compaction (see KAFKA-5316), so we allow some extra space when reading
+      // the batches. Checking the message size is only used as a way to check corruption before checking the CRC
+      val batches = log.batches(maxMessageSize * 2).asScala
```
I am wondering if it's really useful to verify the message size during recovery. We already verified the size when the data is appended in the first place. The only possibility for the size to increase is because of cleaning, which legitimate. Perhaps we could just not check the message size here at all? Alternatively, perhaps we can only check the size for non-compacted topics?

hachikuji replied:
I thought this was a safety measure to prevent an OOM if the end of the log somehow got corrupted. I'm not sure what scenarios were in mind when the check was added, but it seems reasonable to preserve it if we can find a suitable upper bound. I'm guessing your discomfort is the fact that the multiplier is chosen somewhat arbitrarily?

ijuma replied:
Yeah, I think it makes sense to keep some bound. Twice the expected size seems safe in that it is both generous while still serving the original purpose.

hachikuji replied:
I talked with Jun offline and the suggestion we had is to use the remaining size of the file as the sanity check instead. If we find a message which claims to occupy more space than is possible, then it is obviously corrupt.

junrao reviewed (commented):
@hachikuji : Good find on max message size during recovery. A couple of comments.

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -139,72 +154,75 @@ private static FilterResult filterTo(Iterable<MutableRecordBatch> batches, Recor
-                    // could incorrectly see an out of sequence error.
-                    builder.overrideLastOffset(batch.lastOffset());
-
+            if (!retainedRecords.isEmpty()) {
```
Hmm, it seems like the optimisation where the batch doesn't change could be important if a slow compression algorithm like gzip is used?

hachikuji replied:
This is a fair point. Let me see what it looks like with my suggestion above. If it's not significantly more complex, maybe we can keep the optimization.

hachikuji committed:
Address review comments

hachikuji committed:
Remove unused batch iterator accepting max record size

hachikuji committed:
Minor variable renaming

hachikuji commented:
@junrao @ijuma I've updated the patch to reclaim the unchanged batch optimization. Please take another look. I've removed the check for max message size during recovery. I found that the batch iterator is already verifying that the batch size is within the limit of the file. If we find corrupted data at the end of the log during recovery, the iterator will return early and the bytes will be truncated.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -150,60 +163,86 @@ private static FilterResult filterTo(Iterable<MutableRecordBatch> batches, Recor
+
+                if (filteredBatchSize > batch.sizeInBytes() && filteredBatchSize > maxRecordBatchSize)
+                    log.warn("Record batch from {} with last offset {} exceeded max record batch size {} after cleaning " +
+                                    "(new size is {}). Consumers from version 0.10.1 and earlier will need to " +
```
will need to => may need to? A consumer may have already consumed that message.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -132,22 +113,60 @@ public MemoryRecordsBuilder(ByteBuffer buffer,
+     * @param producerEpoch The epoch of the producer
+     * @param baseSequence The sequence number of the first record in this set
+     * @param isTransactional Whether or not the records are part of a transaction
+     * @param writeLimit The desired limit on the total bytes for this record set (note that this can be exceeded
```
There are a few missing params in the comment.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -736,4 +759,12 @@ public long producerId() {
         return this.producerEpoch;
     }
+
+    public static int minBufferSize(byte magic) {
```
This seems no longer used?

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -264,7 +264,7 @@ class LogSegment(val log: FileRecords,
     maxTimestampSoFar = RecordBatch.NO_TIMESTAMP
     try {
-      for (batch <- log.batches(maxMessageSize).asScala) {
+      for (batch <- log.batches.asScala) {
```
Given this, it seems that we don't need to pass in maxMessageSize to LogSegment.recover()?

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -150,60 +163,86 @@ private static FilterResult filterTo(Iterable<MutableRecordBatch> batches, Recor
+            // early to avoid the need for additional allocations.
+            ByteBuffer outputBuffer = bufferOutputStream.buffer();
+            if (outputBuffer != destinationBuffer) {
+                destinationBuffer.position(initialOutputPosition);
```
Do we still need this? It seems that ByteBufferOutputStream already does this?

junrao commented on clients/src/main/java/org/apache/kafka/common/utils/ByteBufferOutputStream.java:
```diff
@@ -40,18 +42,49 @@ public void write(int b) {
         temp.put(buffer.array(), buffer.arrayOffset(), buffer.position());
+
+        // reset the old buffer's position so that
+        buffer.position(initialPosition);
```
This seems to be an existing issue. In line 84, if seems that we need to first check buffer.hasArray() before using buffer.array()?

hachikuji replied:
Embarrassed I didn't catch that since I thought about it when implementing `write(ByteBuffer)`.

junrao reviewed (commented):
@hachikuji : Thanks for the patch. A few more comments.

Also, in ProducerBatch.createBatchOffAccumulator(), we have the following code.

```
        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer,
                                                             compressionType,
                                                             TimestampType.CREATE_TIME,
                                                             batchSize);
```
It seems that we are passing in batchSize for baseOffset.

hachikuji committed:
Fix invalid base offset and unpreserved fields when splitting batches

hachikuji committed:
Address remaining review comments

hachikuji committed:
Minor rename and remove unused variable

hachikuji commented:
@junrao Good catch in `RecordAccumulator`. I noticed also that batch splitting did not preserve the magic value or the transactional flag. I've pushed a basic fix and I'll open a JIRA for some additional test cases. I've addressed your other comments as well.

hachikuji committed:
Make match failure comment in MockClient a little clearer

junrao commented:
@hachikuji : Thanks for the latest patch. LGTM. I will let you merge it.

hachikuji commented:
Thanks. I've merged to trunk and 0.11.0. I will work on a port for 0.10.2 and submit a separate PR.
